openforcefield/topology/molecule.py

```diff
@@ -1815,26 +1815,39 @@ def to_smiles(self, toolkit_registry=GLOBAL_TOOLKIT_REGISTRY):
         >>> smiles = molecule.to_smiles()
 
         """
-        smiles = self._cached_smiles
-
-        if smiles is not None:
-            return smiles
+        # Initialize cached_smiles dict for this molecule if none exists
+        if self._cached_smiles is None:
+            self._cached_smiles = {}
 
+        # Figure out which toolkit should be used to create the SMILES
         if isinstance(toolkit_registry, ToolkitRegistry):
-            smiles = toolkit_registry.call('to_smiles', self)
+            to_smiles_method = toolkit_registry.resolve('to_smiles')
         elif isinstance(toolkit_registry, ToolkitWrapper):
-            toolkit = toolkit_registry
-            smiles = toolkit.to_smiles(self)
+            to_smiles_method = toolkit_registry.to_smiles
         else:
             raise Exception(
                 'Invalid toolkit_registry passed to to_smiles. Expected ToolkitRegistry or ToolkitWrapper. Got  {}'
                 .format(type(toolkit_registry)))
 
-        self._cached_smiles = smiles
-        return smiles
+        # Get a string representation of the function containing the toolkit name so we can check
+        # if a SMILES was already cached for this molecule. This will return, for example
+        # "RDKitToolkitWrapper.to_smiles"
+        func_qualname = to_smiles_method.__qualname__
+
+        # Check to see if a SMILES for this molecule was already cached using this method
+        if func_qualname in self._cached_smiles.keys():
```

Comment:
I'd remove the `keys()` with the `in` operator.

Reply:
Oh man, I totally thought I'd read some python docs that indicated that `in` checks both keys and values, and I was going to be all like "I know more than Andrea about one thing!". 

Except now I can't find it, and the Python docs unambiguously agree with you. It must have been a dream :-/

Reverting...